### PR TITLE
402 api gateway

### PR DESCRIPTION
**Pull request by dominic22**
add kong gateway

### CODE REVIEW

File: api/Dockerfile-audit-trail-gw [hunk 1]
```diff
@@ -4,7 +4,9 @@ WORKDIR /usr/src/app
 COPY . .
 
 RUN cd shared-modules && npm install && npm run build
-# TODO adjust
+
+WORKDIR /usr/src/app/audit-trail-gw
```
Comment (mastrogiovanni) [inline api/Dockerfile-audit-trail-gw:8]: The best practice require to use always the same directory so /usr/src/app is ok for both services

File: api/Dockerfile-ssi-bridge [hunk 2]
```diff
@@ -4,7 +4,9 @@ WORKDIR /usr/src/app
 COPY . .
 
 RUN cd shared-modules && npm install && npm run build
-# TODO adjust
+
+WORKDIR /usr/src/app/ssi-bridge
```
Comment (mastrogiovanni) [inline api/Dockerfile-ssi-bridge:8]: The best practice require to use always the same directory so /usr/src/app is ok for both services

File: api/audit-trail-gw/src/services/configuration-service.ts [hunk 3]
```diff
@@ -36,7 +36,7 @@ export class ConfigurationService {
 	};
 
 	config: Config = {
-		port: process.env.PORT ? Number.parseInt(process.env.PORT, 10) : 3000,
+		port: process.env.PORT ? Number.parseInt(process.env.PORT, 10) : 3002,
```
Comment (mastrogiovanni) [inline api/audit-trail-gw/src/services/configuration-service.ts:39]: Default must be always the same in all services (so 3000).
Port mapping must be done outside, either in env variable (if you run with npm run) or via docker-compose mapping or via kubernetes services.
There is no need to have two different ports for the two services.

File: api/docker-compose.yml [hunk 4]
```diff
@@ -44,5 +45,23 @@ services:
...
+      IOTA_HORNET_NODE: ${IOTA_HORNET_NODE}
+      IOTA_PERMA_NODE: ${IOTA_PERMA_NODE}
+      DATABASE_URL: ${DATABASE_URL}
+      SERVER_SECRET: ${SERVER_SECRET}
+    restart: always
+    depends_on: 
+      - mongo
+      - initialize-job
+    ports:
+      - 3001:3001
```
Comment (mastrogiovanni) [inline api/docker-compose.yml:64]: 3001:3000

File: kubernetes/audit-trail-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            requests:
+              memory: "100Mi"
+              cpu: "250m"
+            limits:
+              memory: "250Mi"
+              cpu: "500m"
+          readinessProbe:
+            httpGet:
+              path: /info
+              port: 3002
```
Comment (mastrogiovanni) [inline kubernetes/audit-trail-deployment.yaml:65]: Leave default one: 3000

File: kubernetes/audit-trail-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                  name: integration-service-api-secrets
+                  key: SERVER_SECRET
+            - name: "API_KEY"
+              valueFrom:
+                secretKeyRef:
+                  name: integration-service-api-secrets
+                  key: API_KEY
+          ports:
+            - protocol: TCP
+              containerPort: 3002
```
Comment (mastrogiovanni) [inline kubernetes/audit-trail-deployment.yaml:104]: Leave default one: 3000

File: kubernetes/is-init-job.yaml [hunk 7]
```diff
@@ -9,8 +9,8 @@ spec:
       restartPolicy: OnFailure
       containers:
         - name: create-root-identity
-          image: iotaledger/integration-services:v0.1.3
-          imagePullPolicy: IfNotPresent
+          image: ssi-bridge
+          imagePullPolicy: Never
```
Comment (mastrogiovanni) [inline kubernetes/is-init-job.yaml:13]: Why Never?

File: kubernetes/is-service.yaml [hunk 8]
```diff
@@ -1,11 +1,23 @@
...
+  name: audit-trail-gw
 spec:
   selector:
-    app: integration-service-api
+    app: audit-trail-gw
   ports:
     - protocol: TCP
-      port: 3000
-      targetPort: 3000  
+      port: 3002
```
Comment (mastrogiovanni) [inline kubernetes/is-service.yaml:10]: Leave default port: 3000

File: kubernetes/is-service.yaml [hunk 9]
```diff
@@ -1,11 +1,23 @@
...
+      targetPort: 3002
```
Comment (mastrogiovanni) [inline kubernetes/is-service.yaml:11]: Leave default port: 3000

File: kubernetes/is-service.yaml [hunk 10]
```diff
@@ -1,11 +1,23 @@
...
+kind: Service
+metadata:
+  name: ssi-bridge
+spec:
+  selector:
+    app: ssi-bridge
+  ports:
+    - protocol: TCP
+      port: 3001
+      targetPort: 3001
```
Comment (mastrogiovanni) [inline kubernetes/is-service.yaml:23]: Leave default port: 3000

File: kubernetes/kong-gw/is-config.yaml [hunk 11]
```diff
@@ -3,8 +3,9 @@ kind: ConfigMap
 metadata:
   name: integration-service-api-config
 data:
-  PORT: "3000"
+  PORT_AUDIT: "3002"
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/is-config.yaml:6]: Port mapping must be managed by the runtime, not the environment variables.
These should be removed

File: kubernetes/kong-gw/is-config.yaml [hunk 12]
```diff
@@ -3,8 +3,9 @@ kind: ConfigMap
...
+  PORT_SSI: "3001"
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/is-config.yaml:7]: Port mapping must be managed by the runtime, not the environment variables.
These should be removed

File: kubernetes/audit-trail-deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+              cpu: "250m"
+            limits:
+              memory: "250Mi"
+              cpu: "500m"
+          readinessProbe:
+            httpGet:
+              path: /info
+              port: 3002
+          env:
+            - name: "PORT"
```
Comment (mastrogiovanni) [inline kubernetes/audit-trail-deployment.yaml:67]: No need for PORT section: container start always on same port. Port mapping must be external

File: kubernetes/kong-gw/kong.yml [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  rules:
+  - http:
+      paths:
+      - path: /api(/|$)(.*)/channel-info
+        pathType: Prefix
+        backend:
+          service:
+            name: audit-trail-gw
+            port:
+              number: 3002
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:18]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+      - path: /api(/|$)(.*)/channels
+        pathType: Prefix
+        backend:
+          service:
+            name: audit-trail-gw
+            port:
+              number: 3002
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:25]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      - path: /api(/|$)(.*)/subscriptions
+        pathType: Prefix
+        backend:
+          service:
+            name: audit-trail-gw
+            port:
+              number: 3002
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:32]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+            name: ssi-bridge
+            port:
+              number: 3001
+      - path: /api(/|$)(.*)/identities
+        pathType: Prefix
+        backend:
+          service:
+            name: ssi-bridge
+            port:
+              number: 3001
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:46]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+            name: audit-trail-gw
+            port:
+              number: 3002
+      - path: /api(/|$)(.*)/authentication
+        pathType: Prefix
+        backend:
+          service:
+            name: ssi-bridge
+            port:
+              number: 3001
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:39]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+      - path: /api(/|$)(.*)/verification
+        pathType: Prefix
+        backend:
+          service:
+            name: ssi-bridge
+            port:
+              number: 3001
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:53]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  rules:
+  - http:
+      paths:
+      - path: /info
+        pathType: Prefix
+        backend:
+          service:
+            name: audit-trail-gw
+            port:
+              number: 3002
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:71]: 3000

File: kubernetes/kong-gw/kong.yml [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+              number: 3001
+      - path: /api(/|$)(.*)/verification
+        pathType: Prefix
+        backend:
+          service:
+            name: ssi-bridge
+            port:
+              number: 3001
+---
+# TODO#415 this should later link to the docs and info of each microservice
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:55]: Are you doing this in another task? If task is already created, please remove the comment

File: kubernetes/kong-gw/kong.yml [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+            name: audit-trail-gw
+            port:
+              number: 3002
+      - path: /docs
+        pathType: Prefix
+        backend:
+          service:
+            name: audit-trail-gw
+            port:
+              number: 3002
```
Comment (mastrogiovanni) [inline kubernetes/kong-gw/kong.yml:78]: 3000

File: kubernetes/ssi-bridge-deployment.yaml [hunk 23]
```diff
@@ -61,15 +62,15 @@ spec:
           readinessProbe:
             httpGet:
               path: /info
-              port: 3000
+              port: 3001
```
Comment (mastrogiovanni) [inline kubernetes/ssi-bridge-deployment.yaml:65]: 3000

File: kubernetes/ssi-bridge-deployment.yaml [hunk 24]
```diff
@@ -61,15 +62,15 @@ spec:
...
           env:
             - name: "SERVER_IDENTITY_FILE"
               value: "/config/server-identity.json"
             - name: "PORT"
```
Comment (mastrogiovanni) [inline kubernetes/ssi-bridge-deployment.yaml:69]: Remove the environment PORT

File: kubernetes/ssi-bridge-deployment.yaml [hunk 25]
```diff
@@ -103,4 +104,4 @@ spec:
                   key: API_KEY
           ports:
             - protocol: TCP
-              containerPort: 3000
+              containerPort: 3001
```
Comment (mastrogiovanni) [inline kubernetes/ssi-bridge-deployment.yaml:107]: 3000

**Review by mastrogiovanni — CHANGES_REQUESTED** [review]
Basically almost all the change request are related to port management.
According the [12 Factor app](https://12factor.net/) port mapping must be done at runtime level.
Containers should not have knowledge about services around them.
For this reason it's ok (and it is a best practice) to leave the default port in runtime (docker-compose and kubernetes) and orchestrate port mapping via Services (in Kubernetes) or Port Mapping (in docker-compose)

File: api/Dockerfile-ssi-bridge (hunk 2 above)
Comment (dominic22) [reply api/Dockerfile-ssi-bridge:8]: But then I am not able to call `RUN npm install` & `RUN npm run build` how would you solve this?

File: api/audit-trail-gw/src/services/configuration-service.ts (hunk 3 above)
Comment (dominic22) [reply api/audit-trail-gw/src/services/configuration-service.ts:39]: 👍

File: kubernetes/ssi-bridge-deployment.yaml (hunk 24 above)
Comment (dominic22) [reply kubernetes/ssi-bridge-deployment.yaml:69]: also for the audit trail then?

File: kubernetes/kong-gw/kong.yml (hunk 21 above)
Comment (dominic22) [reply kubernetes/kong-gw/kong.yml:55]: Yes task is #415 as linked in the todo just marked it as todo so we are aware this needs to be changed

**Review by mastrogiovanni — APPROVED** [review]

**Pull request merged**